tools/codegen/api/python.py

```diff
@@ -0,0 +1,987 @@
+    #
+    # It has 1-1 mapping with DispatchLambdaArgument.
+    exprs: Sequence[str]
+
+@dataclass(frozen=True)
+class LocalInitExprs:
+    # The exprs for special local inits, e.g.:
+    #
+    #   'auto out = _r.tensorlist_n<2>(2);'
+    #
+    exprs: Sequence[str]
+
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ #
+#
+#                          Helper Functions
+#
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ #
+
+# The original simple_type is derived from the 'type' field in Declaration.yaml,
+# which is generated from the C++ argument type, following some seemingly
+# artificial rules:
+#
+# Concrete C++ types are preferred in most cases, e.g.:
+#   'IntArrayRef' instead of 'int[]'
+#   'int64_t' instead of 'int'
+#
+# Constant/Reference annotation and optional field are handled specially, e.g.:
+#   'ScalarType?' instead of 'c10::optional<ScalarType>'
+#   'Tensor' instead of 'const Tensor &' / 'Tensor &'
+#
+# TODO: This needs to be consistent with python_arg_parser - can we simplify it?
+def _simple_type(cpp_type_str: str) -> str:
+    simple_type = cpp_type_str.replace(' &', '').replace('const ', '')
+    opt_match = re.match(r'c10::optional<(.+)>', simple_type)
+    if opt_match:
+        typename = opt_match.group(1)
+        # HACK: 'Layout?' needs to be hardcoded to 'Layout'!
+        simple_type = f'{typename}?' if typename != 'Layout' else 'Layout'
+    return simple_type
+
+def _cpp_signature(f: NativeFunction, *, method: bool = False) -> cpp.CppSignature:
+    return CppSignatureGroup.from_schema(f.func, method=method).signature
+
+def has_tensor_options(f: NativeFunction) -> bool:
+    return any(filter(lambda a: isinstance(a, TensorOptionsArguments),
+                      cpp.group_arguments(f.func, method=False)))
+
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ #
+#
+#                          Python Signature
+#
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ #
+
+def argument(cpp_arg: CppArgument) -> PythonArgument:
+    a = cpp_arg.argument
+    if not isinstance(a, Argument):
+        # cpp's TensorOptionsArguments is ignored, we will reintroduce the
+        # scattered fields in tensor_options_args.
+        raise RuntimeError(f'unsupported cpp argument: \'{cpp_arg}\'')
+    return PythonArgument(
+        name=a.name,
+        type=a.type,
+        cpp_type_str=cpp_arg.type,
+        default=str(pythonify_default(cpp.default_expr(a.default, a.type)))
+        if a.default is not None else None,
+        default_init=None,
+    )
+
+def signature(f: NativeFunction, *, method: bool = False) -> PythonSignature:
+    # Use cpp api to gather TensorOptions fields from kwargs.
+    # Always set 'method' to false as ThisArgument is not relevant - 'self'
+    # is still included as regular Argument type.
+    # TODO: maybe directly generate from FunctionSchema to avoid slicing back
+    # into args/kwargs/outputs?
```

Comment:
I agree with the spirit of this TODO. The awkwardness, I suppose, is that you still need the CppArguments in some cases. Need to study a bit more what exactly these are going to get used for.